bpf/bpf_lxc.c

```diff
@@ -774,6 +798,10 @@ int tail_handle_ipv4(struct __ctx_buff *ctx)
 		return send_drop_notify(ctx, SECLABEL, dstID, 0, ret,
 					CTX_ACT_DROP, METRIC_EGRESS);
 
+	if (!encode_custom_prog_meta(ctx, ret, dstID))
+		tail_call_static(ctx, &CUSTOM_CALLS_MAP,
+				 CUSTOM_CALLS_IDX_IPV4_EGRESS);
```

Comment:
Should we still increment some metric if the tail call fails?

Reply:
Why not, that would make sense. Do you think I should `DROP_MISSED_TAIL_CALL`, or add new dedicated metrics?

Reply:
[Unresolving] So I went from `DROP_MISSED_TAIL_CALL`, but I'm reverting this now, this is not what we want. We do expect a lot of those tail calls to not happen, because the hook is active as soon as we enable the support for custom programs in the agent but it will effectively be taken only if the user has set up a custom program to be called in the prog_array map.

So if we want to know how many times the tail call “fails” (or should we say, does not occur), we would need another metric. But I'm not sure how relevant this would be, because you could have “missed” tail calls just in the duration between the LXC program is attached, and the prog_array map for the endpoint is updated from the user, and this should not necessarily be seen as a bug or anything.

Reply:
If we assume that the custom progs are dynamically added & removed at any point during runtime then yeah I'm not sure what use it will be to track these. I can imagine coming up with some solution like dynamically enabling such a metric programmatically from Cilium via static data + reload or something but I feel like that's overly complicated and the benefit seems low.

At any given point in time, we can dump the custom tail calls prog array map and see whether the prog is there and infer whether it's being called or not. (Which reminds me, we should add this optional map dump to the debuginfo...)

Reply:
A metric is not an error or a warning. It's up to the user to interpret them. We already report a lot of metrics (e.g., packet drops, regenerating endpoints, etc.) which can, under the right circumstances, be interpreted as either benign counters or errors. In addition, I'd expect most users to collect metrics over time and watch their evolution rather than their absolute value.

Right now, if a user suspects the tail call is missed, they have no clear way to validate. They have to implement some sort of logging in their own program and trust it to be correct. Not the most pain-free way to start extending Cilium's datapath IMO.

Reply:
You have a point. Do you mind if I open a follow-up issue once this is merged so we can come back on this? I'd like to avoid adding cycles on implementation, naming, and testing of the metrics in this PR.

One note on metrics evolution: In this case, the metrics will likely always be increasing when hooks for custom programs are enabled, unless _all_ hooks have a destination program attached (or are unused in the datapath), so this may be harder to interpret. But as you say, it's probably better than no logging.

Reply:
> Do you mind if I open a follow-up issue once this is merged so we can come back on this?

Sure :+1: Let's just :crossed_fingers: that it doesn't introduce a complexity issue.